torch/autograd/gradcheck.py

```diff
@@ -867,36 +870,46 @@ def fast_gradcheck(func, func_out, tupled_inputs, outputs, eps, rtol,
 
     # Use our own generator to avoid messing with the user's RNG state
     g_cpu = torch.Generator()
-    all_u = [vec_from_tensor(inp, g_cpu) for inp in inp_tensors]
+    all_ur = [vec_from_tensor(inp, g_cpu, True) for inp in inp_tensors]
+    all_ui = [vec_from_tensor(inp, g_cpu, True) for inp in inp_tensors]
     all_v = [vec_from_tensor(out, g_cpu) for out in outputs]
 
     if not outputs:
-        check_no_differentiable_outputs_fast(func, func_out, tupled_inputs, inp_tensor_indices, all_u, eps, nondet_tol)
+        check_no_differentiable_outputs_fast(func, func_out, tupled_inputs, inp_tensor_indices,
+                                             all_ur, all_ui, eps, nondet_tol)
 
     # Initialize list of lists to store jacobians for each input, output pair
     all_analytical: List[List[torch.Tensor]] = [[] for _ in outputs]
     all_numerical: List[List[torch.Tensor]] = [[] for _ in inp_tensors]
 
     # Numerically approximate v^T (J u)
-    for i, (inp, input_idx, u) in enumerate(zip(inp_tensors, inp_tensor_indices, all_u)):
-        numerical = get_jvp_wrt_specific_input(func, input_idx, tupled_inputs, outputs, u, eps)
+    for i, (input_idx, ur, ui) in enumerate(zip(inp_tensor_indices, all_ur, all_ui)):
+        numerical = get_jvp_wrt_specific_input(func, input_idx, tupled_inputs, outputs, ur, ui, eps)
         for j, (a, v) in enumerate(zip(numerical, all_v)):
             all_numerical[i].append(dot_with_type_promotion(a, v))
 
     # Analytically calculate (v^T J) u
-    all_u_dense = [u.to_dense().reshape(-1) if u.layout == torch.sparse_coo else u for u in all_u]
+    all_ur_dense = [u.to_dense().reshape(-1) if u.layout == torch.sparse_coo else u for u in all_ur]
+    all_ui_dense = [u.to_dense().reshape(-1) if u.layout == torch.sparse_coo else u for u in all_ui]
     for i, (out, v) in enumerate(zip(outputs, all_v)):
         analytical = check_analytical_jacobian_attributes(tupled_inputs, out, nondet_tol, check_grad_dtypes,
                                                           fast_mode=True, v=v)
-        for a, u in zip(analytical, all_u_dense):
-            all_analytical[i].append(a.T.squeeze(0).dot(u))
+        for a, ur, ui in zip(analytical, all_ur_dense, all_ui_dense):
+            if a.is_complex():
+                av = torch.view_as_real(a.T.squeeze(0))
+                ar = av.select(-1, 0)
+                ai = av.select(-1, 1)
+                all_analytical[i].append(ar.dot(ur) + 1j * ai.dot(ui))
+            else:
+                all_analytical[i].append(dot_with_type_promotion(a.T.squeeze(0), ur))
```

Comment:
why do you need dot_with_type_promotion for the real case but it wasn't needed before?

Reply:
This is not needed actually, this was resolved by the "complex downcasting" suggestion above